FF116 Intl.PluralRules - NumFormat v3 (#28097)
mdn/content: files/en-us/web/javascript/reference/global_objects/intl/pluralrules/index.md

The **`Intl.PluralRules`** object enables plural-sensitive formatting and plural-related language rules.

> ## Concepts and usage

Reviewer: ```suggestion
## Description
```

Author: This is what the Web APIs to for an API overview. I can't find a good equivalent example in JavaScript - the overview pattern seems to be to just jump into Examples. I assume this is the pattern you want though, so thanks.

Languages use different patterns for expressing both plural numbers of items (cardinal numbers) and for expressing the order of items (ordinal numbers).
English has two forms for expressing cardinal numbers: one for the singular "item" (1 hour, 1 dog, 1 fish) and the other for zero or any other number of "items" (0 hours, 2 lemmings, 100000.5 fish), while Chinese has only one form, and Arabic has six!

> Similarly, English has four forms for ordinal numbers: "th", "st", "nd", "rd", giving the sequence: 0th, 1st, 2nd, 3rd, 4th, 5th, ..., 21st, 22nd, 23rd, 24th, 25th, and so on, while both Chinese and Arabic only have one form for ordinal numbers.

Reviewer: ```suggestion
Similarly, English has four forms for expressing ordinal numbers: "th", "st", "nd", "rd", giving the sequence: 0th, 1st, 2nd, 3rd, 4th, 5th, ..., 21st, 22nd, 23rd, 24th, 25th, and so on, while both Chinese and Arabic only have one form for ordinal numbers.
```

> The plural rules methods [`PluralRules.select()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/select) and [`PluralRules.selectRange()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/selectRange) return a _tag_ that represents the plural form of a particular cardinal or ordinal number, or range, for a particular language and set of formatting options.

Reviewer: ```suggestion
Given a particular language and set of formatting options, the methods [`Intl.PluralRules.prototype.select()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/select) and [`Intl.PluralRules.prototype.selectRange()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/selectRange) return a _tag_ that represents the plural form of a single or a range of numbers, cardinal or ordinal.
```

> Code can use the returned tags to format strings appropriately for the given language and number/range.

Reviewer: ```suggestion
Code can use the returned tags to represent numbers appropriately for the given language.
```

> The full set of tags that might be returned are: `zero`, `one`, `two`, `few`, `many`, `other` (the "general" plural form, used if the language only has one form).

Reviewer: ```suggestion
The full set of tags that might be returned are: `zero`, `one`, `two`, `few`, `many`, and `other` (the "general" plural form, used if the language only has one form).
```

Also, if English returns `one`/`other`, that already contradicts the statement here. I think you mean "_also_ used if the language only has one form".

Author: I didn't actually think about it, but you're right - that is what I should have meant :-0. Fixed.

> As English only has two forms for cardinal numbers, the `PluralRules.select()` method return only two tags: `"one"` for the singular case, and `"other"` for all other cardinal numbers.

Reviewer: ```suggestion
As English only has two forms for cardinal numbers, the `select()` method returns only two tags: `"one"` for the singular case, and `"other"` for all other cardinal numbers.
```

> This allows constructions of sentences that make sense in English: "1 dog _is_ happy" and "10 dogs _are_ happy", or "1 fish _is_ golden" and "10 fish _are_ golden".

Reviewer: This is very interesting, because when people think of "plural" they usually think of changing the noun itself. I think you would want to point out that there are three kinds of variation in pluralization:

1. Inflection of the noun itself: dog–dogs
2. Inflection of the verb (if there's agreement in pluralization): is–are
3. Inflection of the referent: it–them

And even if the noun itself never changes, the other two variations would also be accounted as a different pluralization. The CLDR spec also talks about this.

> `PluralRules.select()` can return any of four tags for ordinal numbers in English, representing each of the allowed forms: `one` for "st" numbers (1, 21, 31, ...), `two` for "nd" numbers (2, 22, 32, ...), `few` for "rd" numbers (3, 33, 43, ...), and `other` for "th" numbers (0, 4-20, etc.).

Reviewer: ```suggestion
`select()` can return any of four tags for ordinal numbers in English, representing each of the allowed forms: `one` for "st" numbers (1, 21, 31, ...), `two` for "nd" numbers (2, 22, 32, ...), `few` for "rd" numbers (3, 33, 43, ...), and `other` for "th" numbers (0, 4-20, etc.).
```

Again, the returned tags allow appropriate formatting of strings describing an ordinal number.

> For more information about the rules and how they are used see [Plural Rules](https://cldr.unicode.org/index/cldr-spec/plural-rules).
> For a list of the rules and how they apply for different languages see the [LDML Language Plural Rules](https://www.unicode.org/cldr/charts/43/supplemental/language_plural_rules.html).

Reviewer: ```suggestion
For more information about the rules and how they are used, see [Plural Rules](https://cldr.unicode.org/index/cldr-spec/plural-rules).
For a list of the rules and how they apply for different languages, see the [LDML Language Plural Rules](https://www.unicode.org/cldr/charts/43/supplemental/language_plural_rules.html).
```

The **`Intl.PluralRules`** object enables plural-sensitive formatting and plural-related language rules.

> ## Concepts and usage

Author: This is VERY high level. What I'm trying to do here is make it clear that the tags are just tags, not "words" you can use for  direct insertion.

Code can use the returned tags to represent numbers appropriately for the given language.
The full set of tags that might be returned are: `zero`, `one`, `two`, `few`, `many`, and `other` (the "general" plural form, also used if the language only has one form).

As English only has two forms for cardinal numbers, the `select()` method returns only two tags: `"one"` for the singular case, and `"other"` for all other cardinal numbers.
This allows construction of sentences that make sense in English for each case, such as: "1 dog is happy; do you want to play with it?" and "10 dogs are happy; do you want to play with them?".

Creating appropriate sentences for each form depends on the language, and even in English may not be as simple as just adding "s" to a noun to make the plural form.
Using the example above, we see that the form may affect:

- **Nouns**: 1 dogs/2 dogs (but not "fish" or "sheep", which have the same singular and plural form).
- **Verbs**: 1 dog _is_ happy, 2 dogs _are_ happy
- **Pronouns** (and other referents): Do you want to play with _it_ / _them_.

> Other languages have more forms, and choosing appropriate sentences can be even more complex.

Author: @Josh-Cena 

With respect to https://github.com/mdn/content/pull/28097#discussion_r1273983337  (" This is very interesting, because when people think of "plural" they usually think of changing the noun ...)

Yes. I've stolen the form of your comment and inserted it in the text here. I had thought about this, but found it hard to decide what should be included, particularly when I don't "grok" the full discussion in https://cldr.unicode.org/index/cldr-spec/plural-rules - especially when it gets down to the Welsh.

Its a good suggestion. Hope this meets with your approval. Other suggestions merged.

Reviewer: Yep, this looks very clear 👍

### Using locales

This example shows some of the variations in localized plural rules for cardinal numbers.

In order to get the format for the language used in the user interface of your application, make sure to specify that language (and possibly some fallback languages) using the [constructor `locales`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/PluralRules#locales) argument:

```js
// US English
const enCardinalRules = new Intl.PluralRules("en-US");

> console.log(enCardinalRules.select(0)); // > "other"
> console.log(enCardinalRules.select(1)); // > "one"
> console.log(enCardinalRules.select(2)); // > "other"
> console.log(enCardinalRules.select(3)); // > "other"
> 
> // Arabic
> const arCardinalRules = new Intl.PluralRules("ar-EG");
> console.log(arCardinalRules.select(0)); // > "zero"
> console.log(arCardinalRules.select(1)); // > "one"
> console.log(arCardinalRules.select(2)); // > "two"
> console.log(arCardinalRules.select(6)); // > "few"
> console.log(arCardinalRules.select(18)); // > "many"

Reviewer: ```suggestion
console.log(enCardinalRules.select(0)); // "other"
console.log(enCardinalRules.select(1)); // "one"
console.log(enCardinalRules.select(2)); // "other"
console.log(enCardinalRules.select(3)); // "other"

// Arabic
const arCardinalRules = new Intl.PluralRules("ar-EG");
console.log(arCardinalRules.select(0)); // "zero"
console.log(arCardinalRules.select(1)); // "one"
console.log(arCardinalRules.select(2)); // "two"
console.log(arCardinalRules.select(6)); // "few"
console.log(arCardinalRules.select(18)); // "many"
```

```

### Using options

The plural form of the specified number may also depend on [constructor `options`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/PluralRules#locales), such as how the number is rounded, and whether it is cardinal or ordinal.

This example shows how you can set the type of rules to "ordinal", and how this affects the form for some numbers in US English.

```js
// US English - ordinal
const enOrdinalRules = new Intl.PluralRules("en-US", { type: "ordinal" });

> console.log(enOrdinalRules.select(0)); // > "other" (0th)
> console.log(enOrdinalRules.select(1)); // > "one"   (1st)
> console.log(enOrdinalRules.select(2)); // > "two"   (2nd)
> console.log(enOrdinalRules.select(3)); // > "few"   (3rd)
> console.log(enOrdinalRules.select(4)); // > "other" (4th)
> console.log(enOrdinalRules.select(21)); // > "one"  (21st)

Reviewer: ```suggestion
console.log(enOrdinalRules.select(0)); // "other" (0th)
console.log(enOrdinalRules.select(1)); // "one"   (1st)
console.log(enOrdinalRules.select(2)); // "two"   (2nd)
console.log(enOrdinalRules.select(3)); // "few"   (3rd)
console.log(enOrdinalRules.select(4)); // "other" (4th)
console.log(enOrdinalRules.select(21)); // "one"  (21st)
```

### Using locales

This example shows some of the variations in localized plural rules for cardinal numbers.

In order to get the format for the language used in the user interface of your application, make sure to specify that language (and possibly some fallback languages) using the [constructor `locales`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/PluralRules#locales) argument:

```js
// US English
const enCardinalRules = new Intl.PluralRules("en-US");
console.log(enCardinalRules.select(0)); // "other"
console.log(enCardinalRules.select(1)); // "one"
console.log(enCardinalRules.select(2)); // "other"
console.log(enCardinalRules.select(3)); // "other"

// Arabic
const arCardinalRules = new Intl.PluralRules("ar-EG");
console.log(arCardinalRules.select(0)); // "zero"
console.log(arCardinalRules.select(1)); // "one"
console.log(arCardinalRules.select(2)); // "two"
console.log(arCardinalRules.select(6)); // "few"
console.log(arCardinalRules.select(18)); // "many"
```

### Using options

> The plural form of the specified number may also depend on [constructor `options`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/PluralRules#locales), such as how the number is rounded, and whether it is cardinal or ordinal.

Reviewer: ```suggestion
The plural form of the specified number may also depend on [constructor `options`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules/PluralRules#options), such as how the number is rounded, and whether it is cardinal or ordinal.
```
